### PR TITLE
Secrets manager and base provider proposal

### PR DESCRIPTION
**Pull request by protatremy**
Proposal to implement Buildbot secrets service:
- secrets manager (get API)
- secret provider base

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/secrets/secret.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        self._value = value
+        self._props = props
+        self._key = key
+
+    @property
+    def source(self):
+        """
+        source of the secret
+        """
+        return self._source
```
Comment (tardyp) [inline master/buildbot/secrets/secret.py]: please try to get 100% coverage. install codecov.io chrome extension to see that this line is not covered

File: master/buildbot/secrets/secret.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#
+# Copyright Buildbot Team Members
+
+class SecretDetails(object):
+    """
+    A SecretDetails object has secrets attributes:
+    - source: provider  to retrieve secret
+    - key: secret key identifier
+    - value: secret value
+    - props: all other needed properties
```
Comment (tardyp) [inline master/buildbot/secrets/secret.py]: what kind of properties do we need for secrets?

File: master/buildbot/secrets/secret.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+class SecretDetails(object):
+    """
+    A SecretDetails object has secrets attributes:
+    - source: provider  to retrieve secret
```
Comment (tardyp) [inline master/buildbot/secrets/secret.py]: provider where the secret was retrieved

Comment (protatremy) [inline master/buildbot/secrets/secret.py]: replaced

File: master/buildbot/secrets/manager.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        get secrets from the provider defined in the secret using args and
+        kwargs
+        @secrets: secrets keys
+        @type: string
+        @return type: SecretDetails
+        """
+        secret_detail = None
+        providers = self.master.config.secretsManagers
+        for provider in providers:
+            value = provider.get(secret)
```
Comment (tardyp) [inline master/buildbot/secrets/manager.py]: I think we need this get to be yield. Some providers will need to make network access to do the get

File: master/buildbot/secrets/manager.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+from buildbot.util import service
+
+
+class SecretManager(service.BuildbotService):
+    """
+    Secret manager
+    """
+    name = 'secrets'
+
+    def get(self, secret, *args, **kwargs):
```
Comment (tardyp) [inline master/buildbot/secrets/manager.py]: this function shall return deferred (inlinecallbacks)

File: master/buildbot/secrets/provider/base.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+class SecretProviderBase(BuildbotService):
+    """
+    ...
+    """
+
+    def get(self, *args, **kwargs):
+        """
+        this should be an abstract method
+        """
```
Comment (tardyp) [inline master/buildbot/secrets/provider/base.py]: the you can mark it abstract https://docs.python.org/2/library/abc.html

File: master/docs/developer/secrets.rst [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+A secret is identified by a couple (key, value).
+
+.. code-block:: python
+
+  class SecretDetails(object):
+
+      def __init__(self, provider, key, value, props=None):
+
+A ``secretDetails`` is a python object initialized with the following parameters:
+- provider name to retrieve secrets,
```
Comment (tardyp) [inline master/docs/developer/secrets.rst]: name of provider where secret has been retrieved

File: master/docs/developer/secrets.rst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+- key identifier
+- value returned by the provider API
+- properties if needed.
+
+Each parameter is an object property that should returned the value.
```
Comment (tardyp) [inline master/docs/developer/secrets.rst]: its not very clear to me what you mean

File: master/docs/developer/secrets.rst [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    secretDetails = secretsService.get(secret)
+
+The get API take the secret key as parameters and read the configuration to obtains the list of configured providers.
+The manager get the selected provider and returns a ``SecretDetails``.
+
+.. code-block:: python
+
+  c['secretsProviders'] = [SecretsProviderOne(params), SecretsProviderTwo(params)]
+
+If more than one provider is defined in the configuration, the manager returns the first founded value.
```
Comment (tardyp) [inline master/docs/developer/secrets.rst]: the manager returns the first value found.

File: master/docs/developer/secrets.rst [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+```````````````
+The secret manager is a Buildbot service, providing a get method API to retrieve a secret value.
+
+.. code-block:: python
+
+    secretsService = self.master.namedServices['secrets']
+    secretDetails = secretsService.get(secret)
+
+The get API take the secret key as parameters and read the configuration to obtains the list of configured providers.
+The manager get the selected provider and returns a ``SecretDetails``.
```
Comment (tardyp) [inline master/docs/developer/secrets.rst]: +The manager call the get method of the configured provider and returns a ``SecretDetails`` if the call succeed.

File: master/buildbot/test/unit/test_fake_secrets_manager.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+    def testGetDataMultipleManagerServiceNoDatas(self):
+        secret_service_manager = SecretManager()
+        self.master.config.secretsManagers = [FakeSecretStorage({"foo": "bar",
+                                                                 "other": "value"}),
+                                              FakeSecretStorage({"foo2": "bar",
+                                                                 "other2": "value"})
+                                              ]
+        SecretManager.master = self.master
+        expectedSecretDetail = SecretDetails(
```
Comment (tardyp) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: If not found, I think it should return None, not secretDefails(fakestorage2, None)

File: master/buildbot/test/unit/test_fake_secrets_manager.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        secret_service_manager = SecretManager()
+        self.master.config.secretsManagers = [FakeSecretStorage({"foo": "bar",
+                                                                 "other": "value"}),
+                                              OtherFakeSecretStorage({"foo2": "bar",
+                                                                      "other2": "value"},
+                                                                     props={"property": "value_prop"})
+                                              ]
+        SecretManager.master = self.master
+        expectedSecretDetail = SecretDetails(
+            OtherFakeSecretStorage({"foo2": "bar",
```
Comment (tardyp) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: why dont you just write ```OtherFakeSecretStorage.__name__ ```?

File: master/buildbot/test/unit/test_fake_secrets_manager.py (hunk 11 above)
Comment (protatremy) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: I preferred to return the object because it contains the provider class name information, that will be used to prompt a log error indication from where the information has not been found.

Comment (protatremy) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: I preferred to return the object because it contains the provider class name information, that will be used to prompt a log error indication from where the information has not been found.

File: master/buildbot/secrets/manager.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        get secrets from the provider defined in the secret using args and
+        kwargs
+        @secrets: secrets keys
+        @type: string
+        @return type: SecretDetails
+        """
+        secret_detail = None
+        providers = self.master.config.secretsManagers
+        for provider in providers:
```
Comment (tardyp) [inline master/buildbot/secrets/manager.py]: you should return the first that is not None.

    for provider in providers:
        value = yield provider.get(secret)
        if value is not None:
            source_name = provider.__class__.__name__
            defer.returnValue(SecretDetails(source_name, secret, value))
    # return None is implicit

File: master/buildbot/test/unit/test_fake_secrets_manager.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+                                                                 "other": "value"}),
+                                              FakeSecretStorage({"foo2": "bar",
+                                                                 "other2": "value"})
+                                              ]
+        SecretManager.master = self.master
+        expectedSecretDetail = SecretDetails(FakeSecretStorage.__name__,
+                                             "foo3",
+                                             None)
+        secret_result = yield secret_service_manager.get("foo3")
+        self.assertEqual(secret_result, expectedSecretDetail)
```
Comment (tardyp) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: Need to add a test for 2 providers having the secret in their DB (but not the same value):
should return the value of the first provider

File: master/buildbot/secrets/provider/__init__.py [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
```
Comment (tardyp) [inline master/buildbot/secrets/provider/__init__.py]: We will have several providers, so we should call this module buildbot.secrets.providers (with an s)

File: master/buildbot/secrets/manager.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        kwargs
+        @secrets: secrets keys
+        @type: string
+        @return type: SecretDetails
+        """
+        providers = self.master.config.secretsManagers
+        for provider in providers:
+            value = yield provider.get(secret)
+            source_name = provider.__class__.__name__
+            if value:
```
Comment (tardyp) [inline master/buildbot/secrets/manager.py]: if value is not None  (a secret might be empty string)

File: master/buildbot/test/unit/test_fake_secrets_manager.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    def testGetDataMultipleManagerServiceNoDatas(self):
+        secret_service_manager = SecretManager()
+        self.master.config.secretsManagers = [FakeSecretStorage({"foo": "bar",
+                                                                 "other": "value"}),
+                                              FakeSecretStorage({"foo2": "bar",
+                                                                 "other2": "value"})
+                                              ]
+        SecretManager.master = self.master
+        secret_result = yield secret_service_manager.get("foo3")
+        self.assertEqual(secret_result, None)
```
Comment (tardyp) [inline master/buildbot/test/unit/test_fake_secrets_manager.py]: I don't see the test with two providers providing the same secret.

Need another test with a secret value being empty string ( empty string is valid secret)
